### ISSUE TITLE
torch.linalg in PyTorch 1.10 tracker

### ISSUE THREAD

**Issue by mattip** [pytorch/pytorch#42666]
This is a tracking issue for torch.linalg tasks for PyTorch 1.10. The goals for the 1.10 release are:

- [Reviewing the Python Array AP's](https://data-apis.org/array-api/latest/extensions/linear_algebra_functions.html) linalg extension and pursuing consistency with it
- Modernizing linear algebra functions still under the torch namespace to newer torch.linalg versions
- Continuing to deprecate older linear algebra operators (see https://github.com/pytorch/pytorch/issues/55953)
- Improving CUDA performance (see https://github.com/pytorch/pytorch/issues/47953)
- Improving linear algebra gradient computations

**Tasks**

- [X] Add the "upper" kwarg to linalg.cholesky #62434
- [ ] Rationalize the kwargs of linalg.eigh, linalg.eigvalsh, linalg.lstsq, and linalg.pinv with each other and the Python Array API
 - https://github.com/pytorch/pytorch/issues/54151#issuecomment-879937591
- [x] Add modern versions of lu, lu_unpack, geqrf, orgqr, ormqr, lu_solve, triangular_solve, cholesky_solve, cholesky_inverse, tensordot, matrix_exp, {svd, pca}_lowrank and lobpcg to torch.linalg
 - https://github.com/pytorch/pytorch/issues/61648
 - https://github.com/pytorch/pytorch/issues/61649
 - https://github.com/pytorch/pytorch/issues/61650
 - https://github.com/pytorch/pytorch/issues/61653
 - https://github.com/pytorch/pytorch/issues/61657
 - https://github.com/pytorch/pytorch/issues/61658
- [ ] Investigate a common "solver API" to rationalize the various solvers

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @jianyuh @nikitaved @pearu @mruberry @heitorschueroff @walterddr @IvanYashchuk @xwang233 @Lezcano @rgommers @vincentqb @vishwakftw @SsnL

**Comment by rgommers** [comment]
xref gh-42053, the correct way to implement aliases is still under discussion there.

**Comment by mruberry** [comment]
> xref [gh-42053](https://github.com/pytorch/pytorch/issues/42053), the correct way to implement aliases is still under discussion there.

We ultimately decided on the approach taken in https://github.com/pytorch/pytorch/pull/42586. In the future we may implement an "alias" entry in native_functions.yaml, too.

**Comment by albanD** [comment]
I think an important thing to keep in mind for these new ops is to make sure we don't do the same mistake as the current ones and provide proper documentation about them.
In particular, it should contain enough information for people that know the field to know what happens (for example, high level functions should mention how they are implemented)
Moreover, for less experience people, it would be important to have informations like:
- What are the requirements on the inputs for this function to be valid
- What are the numerical properties of this function in general
- When it exists, propose a more numerically stable alternative for the problematic inputs

Finally, when the backward is not implemented for all cases that the function supports, this should be mentioned as well.

**Comment by rgommers** [comment]
Regarding the task list here: I'm not too sure about deprecating so many functions. For `norm` it makes sense, because the `torch.linalg.norm` function is different from (and arguably nicer) than `torch.norm`. But for simple aliases, I'd at least do those last, because there's only value once the whole `linalg` namespace is complete so we can say "`torch.linalg` matches `numpy.linalg`". And for some heavily used functions I'd consider not doing the deprecation at all.

**Comment by rgommers** [comment]
@mruberry, @ezyang and I had a chat about priorities:

1. Implement functions that don’t yet exist, most-used ones first (e.g. `kron`, `eigh` - here is some [usage data](https://github.com/Quansight-Labs/python-api-inspect/blob/master/data/csv/numpy-summary.csv))
2. Implement functions that exist in `torch` but need a change to match numpy when they go into `torch.linalg`
3. Add aliases to linalg namespace for functions that exist and already match numpy
4. Add the necessary deprecations to old functions in one go when the `torch.linalg` namespace is complete (TBD whether they should all be deprecated, see my comment above).

**Comment by mruberry** [comment]
@rgommers I updated the rollup to reflect these priorities and remove all deprecation tasks (except deprecating torch.norm, as we discussed earlier). I appreciate your point about maybe being less aggressive with deprecations than this initially was. Let's review on a case-by-case basis.

**Comment by antocuni** [comment]
picking up `torch.linalg.svd`

**Comment by IvanYashchuk** [comment]
Picking up `kron`, `eigh`, `cond`, `tensorinv`.

**Comment by antocuni** [comment]
I have a design question about `torch.linalg.svd`, but I think it's general enough that it's good to discuss it here.
`numpy.lingalg.svd` differs from `torch.svd` in two major ways:
1. the numpy version has a parameter called `full_matrices`, while `torch.svd` has one called `some` which is basically the opposite, so it's easy to adapt

2. the numpy signature is irregular, and the return type changes depending on the value of `compute_uv`. In particular, if `compute_uv==True` it returns a 3-tuple, else it returns a single array. OTOH, `torch.svd` always return a 3-tuple.

I see that the the existing `torch.linalg.*` functions all provide a C++ API in the `torch::linalg` namespace, but point 2 makes it hard/impossible to provide a C++ signature which is consistent with the Python one. I see at least two solutions:

1. implement `torch.linalg.svd` entirely in Python on top of `torch.svd`, and not offering the C++ API

2. implement a C++ version of `torch::linalg::svd` which always returns a 3-tuple, and write a thin Python wrapper which discards the two extra elements in the `compute_uv==False` case.

Moreover, what are we going to do with `torch.svd`? Should we:
1. keep it as is and implement `torch.linalg.svd` on top of it OR
2. move the current implementation to `torch.linalg.svd` and re-implement the old `torch.svd` on top of it for backwards compatibility?

**Comment by IvanYashchuk** [comment]
I'm in favor of this solution:
>implement a C++ version of torch::linalg::svd which always returns a 3-tuple, and write a thin Python wrapper which discards the two extra elements in the compute_uv==False case.

I have done the same for ChainerX [here](https://github.com/IvanYashchuk/chainer/blob/a326597e8b731d8cfbf27d44f0981b9ae62e5161/chainerx_cc/chainerx/python/routines.cc#L359-L362).

**Comment by rgommers** [comment]
> Moreover, what are we going to do with `torch.svd`? Should we:
> 
> 1. keep it as is and implement `torch.linalg.svd` on top of it OR
> 2. move the current implementation to `torch.linalg.svd` and re-implement the old `torch.svd` on top of it for backwards compatibility?

It seems to me that (1) is a little less work, but would be much less appealing if `torch.svd` gets deprecated in the future. That's not decided, but seems fairly likely to happen at some point. So (2) is probably preferred. @mruberry does that make sense to you?

**Comment by antocuni** [comment]
> It seems to me that (1) is a little less work, but would be much less appealing if torch.svd gets deprecated in the future. That's not decided, but seems fairly likely to happen at some point. So (2) is probably preferred. @mruberry does that make sense to you?

I opened PR #45562 to discuss this more concretely. I'll wait for some feedback on it before continuing

**Comment by mruberry** [comment]
@antocuni Sorry for the delay, working through a backlog now that the 1.7 branch is cut. I'll get to this ASAP (definitely before Monday).

**Comment by mruberry** [comment]
> > Moreover, what are we going to do with `torch.svd`? Should we:
> > 
> > 1. keep it as is and implement `torch.linalg.svd` on top of it OR
> > 2. move the current implementation to `torch.linalg.svd` and re-implement the old `torch.svd` on top of it for backwards compatibility?
> 
> It seems to me that (1) is a little less work, but would be much less appealing if `torch.svd` gets deprecated in the future. That's not decided, but seems fairly likely to happen at some point. So (2) is probably preferred. @mruberry does that make sense to you?

Unfortunately I'm not very familiar with the code. I would do the simplest thing to implement `torch.linalg.svd()`, as discussed on the PR. We can always rearrange the functions internally later.

**Comment by mruberry** [comment]
@muthuArivoli It's taken awhile, but I was finally able to review linear algebra related issues. You mentioned you were interested in working on one. Would torch.inner(), torch.kron(), fixing torch.einsum, torch.linalg.cond(), or torch.linalg.matrix_rank() be interesting? Or maybe reviewing if there's an alternative to a MAGMA-based implementation of a function like torch.svd() or torch.lstsq?

**Comment by xwang233** [comment]
Thanks for the linear algebra tracking issue. Now that cusolver has been included in pytorch nightly build #42403, and cusolver matrix inverse seems to be better than MAGMA on single matrix case. I would also suggest that we start planning to add cusolver implementations for those linalg operations that are slow on MAGMA. A heuristic to dispatch cusolver (single matrix)/cublas batched/MAGMA would also be much better.

**Comment by xwang233** [comment]
@antocuni Do you have any plan to add cusolver SVD implementations in your PR #45562? Or is anyone working on that or interested in doing so?

**Comment by antocuni** [comment]
> @antocuni Do you have any plan to add cusolver SVD implementations in your PR #45562? 

I think that the PR has already a broader scope than originally planned (because now it fixes also #45821), so it's probably better to avoid adding too much stuff at it and add cusolver in a separate PR (moreover, I have 0 experience with cusolver, so if anyone wants to take it, I'm fine with it).

**Comment by IvanYashchuk** [comment]
I think nobody is currently working on cusolver implementations. I am interested in doing that, we can work on this together @xwang233. We should thoroughly review the performance of magma and cusolver, first. Also I think it's useful not to forget about ROCm here and think how rocsolver fits in this work.

**Comment by mruberry** [comment]
After talking with @ptrblck, we decided to create separate tasks to review performance and look at switching some function to cuSolver. It's a great idea, @xwang233.

**Comment by mruberry** [comment]
@ptrblck, @xwang233 

I had a chance to look at outstanding linear algebra PRs and a few ops. Since we have PRs updating the UX of [SVD](https://github.com/pytorch/pytorch/pull/45562/files) and [Cholesky](https://github.com/pytorch/pytorch/pull/46083), what would you think about seeing if qr, lu, or solve could be switched from MAGMA to cuSolver? 

When switching an op we should look at:

- performance
- numerical accuracy
- amount of cross-device data movement required

**Comment by xwang233** [comment]
Thanks @mruberry . I would prefer to prioritize cusolver `svd` since it's having some bad performance with MAGMA. I haven't check performance or related issues about `cholesky` or others. 

I agree with your idea that we need to check performance and numerical accuracy when adding cusolver implementations.

For cusolver, the CPU-GPU data sync is mostly the `info` check. Usually `info` is calculated on GPU, but has to be synchronously copied to CPU for the check. This may cause performance drop. I suggest that we add an optional argument, or global linear algebra flag, that allows us to **not** check the `info`. With that, basically all cusolver/cublas-based linear algebra can be fully async, since there is no memcpy or mem allocation/free.

**Comment by IvanYashchuk** [comment]
MAGMA implementations are generally hybrid CPU/GPU. They have several variants of functions distinguished by a suffix, the suffix indicates in what memory the input data starts and ends. In the case of SVD, there are no implementations with the `_gpu` suffix. It's not clear whether they are going to implement this ([link](https://icl.cs.utk.edu/magma/forum/viewtopic.php?t=3807)). [There are several algorithms for computing SVD](https://software.intel.com/content/www/us/en/develop/documentation/mkl-developer-reference-c/top/lapack-routines/lapack-least-squares-and-eigenvalue-problem-routines/lapack-least-squares-and-eigenvalue-problem-driver-routines/singular-value-decomposition-lapack-driver-routines.html#singular-value-decomposition-lapack-driver-routines_TBL4-13). Most commonly implemented are `gesvd` (based on QR iterations) and `gesdd` (based on divide-and-conquer). `gesdd` is usually much faster algorithm ([benchmark link](http://www.netlib.org/lapack/lug/node71.html) and used in NumPy, SciPy and here in PyTorch both for CPU and GPU.

cuSOLVER doesn't have `gesdd`, it implements `gesvd`, `gesvdj` (based on Jacobi iterations), `gesvdp` (based on polar decomposition). Here are slides with a few comparisons on cuSOLVER's `gesvdj` vs MKL's `gesvd` (sadly not comparing to MKL's other routines for SVD).

Note that in Tensorflow cuSOLVER is used for SVD and people report that it is slow (the issue is old though and things may have changed https://github.com/tensorflow/tensorflow/issues/13603) and (old) benchmarks show that PyTorch (MAGMA) is much faster https://github.com/yaroslavvb/stuff/tree/master/linalg-benchmark

All that just tells us that we really need thorough benchmarking.

**Comment by mruberry** [comment]
> Thanks @mruberry . I would prefer to prioritize cusolver `svd` since it's having some bad performance with MAGMA. I haven't check performance or related issues about `cholesky` or others. If there are many performance/numerical accuracy reports for other operators, we can prioritize them as well.

Starting with SVD sounds OK. The UX work going on currently with SVD probably won't affect the performance investigation.

We are interested in improving the performance and stability of Cholesky. @heitorschueroff is going to start investigating workarounds to MAGMA's segfaults, but I imagine that any workaround will negatively impact MAGMA performance.

> For cusolver, the only CPU-GPU data sync that I'm aware of is the `info` check. Usually `info` is calculated on GPU, but has to be synchronously copied to CPU for the check. This may cause performance drop. I suggest that we add an optional argument, or global linear algebra flag, that allows us to **not** check the `info`. With that, basically all cusolver/cublas-based linear algebra can be fully async, since there is no memcpy or mem allocation/free.

No global flags please. Either returning the info as a device tensor or adding a flag to check it sounds reasonable. I'm not sure which approach we'd like to pursue but I don't think this choice will block your investigation? One issue with a flag is that it seems like we'd want different defaults for CPU and CUDA inputs, which is a little weird. On the other hand, returning an extra tensor is awkward.

**Comment by antocuni** [comment]
picking up `torch.linalg.qr`

**Comment by antocuni** [comment]
picking "Add torch.eig complex forward (CPU, CUDA)", a WIP PR will be opened soon

**Comment by IvanYashchuk** [comment]
>  Add torch.einsum complex forward (CUDA)
>  torch.einsum complex backward (see complex autograd guide below)

Since `torch.einsum` depends on `sum` and `bmm`, complex input on CUDA is supported with https://github.com/pytorch/pytorch/pull/42553. Complex backward for `bmm` works with https://github.com/pytorch/pytorch/pull/45737.

**Comment by mruberry** [comment]
> > Add torch.einsum complex forward (CUDA)
> > torch.einsum complex backward (see complex autograd guide below)
> 
> Since `torch.einsum` depends on `sum` and `bmm`, complex input on CUDA is supported with #42553. Complex backward for `bmm` works with #45737.

Have we updated the relevant tests for einsum, too?

**Comment by IvanYashchuk** [comment]
> > > Add torch.einsum complex forward (CUDA)
> > > torch.einsum complex backward (see complex autograd guide below)
> > 
> > 
> > Since `torch.einsum` depends on `sum` and `bmm`, complex input on CUDA is supported with #42553. Complex backward for `bmm` works with #45737.
> 
> Have we updated the relevant tests for einsum, too?

Tested locally that with this change https://github.com/pytorch/pytorch/pull/49189/files#r542826413 tests pass.

**Comment by mruberry** [comment]
Thank you @IvanYashchuk! cc @heitorschueroff. Let's be sure to get complex testing for einsum in properly. If not part of that PR then a follow-up. I'll update the tasks to reflect that testing is needed.

**Comment by heitorschueroff** [comment]
I added the complex tests to einsum.

**Comment by IvanYashchuk** [comment]
Picking `torch.linalg.eig/eigvals`

**Comment by antocuni** [comment]
picking `torch.linalg.matrix_power`

**Comment by nikitaved** [comment]
@antocuni , it is being done in https://github.com/pytorch/pytorch/pull/52608.

**Comment by IvanYashchuk** [comment]
> picking `torch.linalg.matrix_power`

@antocuni, there is submitted PR for that https://github.com/pytorch/pytorch/pull/52608

**Comment by antocuni** [comment]
ouch, thanks for pointing it out. Picking ` torch.linalg.det` instead

**Comment by mruberry** [comment]
Closing in favor of individual issues
